## Getting started
  
git clone https://github.com/zhuqianhua/qTar.git <br>
cd qTar <br>
make<br>
cd example<br>
sh run.sh<br>
  
## Introduction

qTar is a software for targetting small rna against a target sequence. It is based on seed <br>
sequence search, combined with SW precise alignment and sequence binding free energy to <br>
rapidly screen potential target genes. It relies on multithreaded execution and <br>
your system needs to support -lpthread. <br>

## Citing


## Reference

[1] Helwak A, Kudla G, Dudnakova T, Tollervey D. Mapping the Human miRNA Interactome  <br>
>by CLASH Reveals Frequent Noncanonical Binding. Cell. 2013;153:654â€“65. pmid:23622248 <br>
[2] miRBase: annotating high confidence microRNAs using deep sequencing data. <br>
>>Kozomara A, Griffiths-Jones S. NAR 2014 42:D68-D73 <br>
[3] Xinbin Dai, Zhaohong Zhuang and Patrick X. Zhao (2018). psRNATarget: a plant  <br>
      small RNA target analysis server (2017 release). Nucleic Acids Research. doi: 10.1093/nar/gky316. <br>
[4] Agarwal V, Bell GW, Nam J, Bartel DP. Predicting effective microRNA target sites in mammalian  <br>
      mRNAs. eLife, 4:e05005 <br>
[5] John B, Enright AJ, Aravin A, Tuschl T, Sander C, Marks DS., PLoS Biol. 2005 Jul;3(7) <br>
[6] Rehmsmeier, Marc and Krueger, Jan RNAhybrid: microRNA target prediction easy, fast  <br>
      and flexible, Nucleic Acids Res. <br>
[7] Wu HJ, Ma YK, Chen T, Wang M, Wang XJ. (2012) PsRobot: a web-based plant small RNA  <br>
      meta-analysis toolbox. Nucleic Acids Res. DOI:10.1093/nar/gks554 <br>
[8] Fahlgren N, Carrington JC (2010) miRNA Target Prediction in Plants. Methods in  <br>
      molecular biology (Clifton, NJ) 592 <br>
[9] Freier et al., Improved free-energy parameters for predictions of RNA duplex stability.  <br>
      Proc. Natl. Acad. Sci. USA 83: 9373-9377 (1986)) <br>
